### PR TITLE
[Ready] Infection Gamemode

### PR DESCRIPTION
**Pull request by Whoneedspacee**
## About The Pull Request

**SPRITES BY THE GOD FURY MCFLURRY**

Adds infection gamemode and the infection commander antagonist. The infection commander is a unique antagonist of which there can only be one of at a time.

The infection commander spawns his core on a starting point with surrounding defenses that slightly encroaches on the stations border walls. He starts with at least 5 other players that control evolving slimes, and the gamemode can only start if there are at least 50 players on the server, as well as 6 eligible antagonists.

There are beacons that he must destroy by attempting to expand with nodes on them, and the player controlled slimes are his main army. The infection can create more slimes from the ghost pool if player controlled human bodies are dragged to the core.

The player controlled slimes do not die but instead take time to respawn if they are killed. They generate their own points as well and can use them to upgrade into different types of slimes while they are dead. They only truly die if the core of the infection is destroyed by legendary weapons.

Legendary weapons are given from doom events which are unique to this gamemode, each event has a unique item with unique effects that are the only thing able to damage the core of the infection.

Legendary weapons mostly exist as a way to prevent the gamemode from ending early, as well as to prevent something like a singularity or a bag of holding from instantly winning the mode. This was quite the issue for blob as a gamemode so this is my attempted solution for it.

Also adds multiple components for usage in the gamemode that could probably be used in many different things outside of this gamemode.

## Why It's Good For The Game

This gamemode is more of a war of attrition type gamemode with both sides gaining and losing power depending on what the players of both sides do. I'm interested in seeing the strategies that players use when they have concrete objectives to defend and attack and multiple of them, unlike nuclear operatives which is more like a wild goose chase around the station.

I don't think this gamemode is that close to being finished, and I'm well open to any changes that are proposed from testing. Even if they are drastic ones that will affect how the entire mode is played.

I'm not sure how this gamemode will play out in this type of game in the first place, and I think I need to do testing in order to move forward with development for this mode, if it's even possible to make it fun.

[Infection v10, most recent version of the gamemode that has been recorded](https://www.youtube.com/watch?v=D6zlblhr1Oc&feature=youtu.be)

## Changelog
:cl: Joe Berry
add: Infection gamemode, a new war-type gamemode that pits players against a constantly expanding and evolving enemy. Defend the beacons and destroy the core to save the station.
imageadd: Sprites by Fury McFlurry
/:cl:

### CODE REVIEW

**Comment by Whoneedspacee** [comment]
I'll once again state since people hate reading walls of text, but I don't think this gamemode is finished. I don't see this as a finished mode and I'm fully welcoming for even major changes, like the core being able to take damage from non-legendary things.

This is meant to be somewhat of my gift to the players of this game for all the amazing things that it's done for me, and I'll try to make it as fun for you guys as possible. Hell even if it isn't fun I've learned so many things from making this utter piece of shit that hopefully I'll be able to do more after I'm done with it for good.

**Comment by wesoda25** [comment]
It seems cool but gamemodes which consist of waiting for an unchanging threat have already been done and become less enjoyable the more one plays them.

**Comment by actioninja** [comment]
le infection has arrived

**Comment by Whoneedspacee** [comment]
It’s a 5 minute waiting time not that long, mostly just for AI/Borg stuff, basic defenses, and all access stuff. Roundtype is announced publicly 30 seconds in.

**Comment by wesoda25** [comment]
I really don’t wanna instantly attack it but publicly announced gamemodes are conceptually bad and go against the nature of tg (paranoia, not knowing what awaits instantly). And its not the waiting period thats bad, but the fact that you know exactly what is awaiting for you in the next round, it gets boring. Its why so many people dislike nukeops and clockcult.

**Comment by Whoneedspacee** [comment]
Fair enough but I don’t really see how it’s much different (at least announcement wise) compared to like wizard summoning ghosts. People seem to have fun with those rounds.

But fair, I’m trying to make the gamemode more “random” with many upgrades, and maybe in the future I’ll make it so the meteor can spawn on any side of the station.

But yeah if the gamemode concept in general sucks, and it tests horribly, I’m fully open to just closing this and shoving some of my components and crap into another PR.

**Comment by lolman360** [comment]
it's finally here

File: code/modules/antagonists/infection/beacon.dm [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+	if(GLOB.infection_beacons.len > 0)
+		INVOKE_ASYNC(src, .proc/destroyed_announcement)
+	return ..()
+
+/*
+	Delayed announcement that occurs when the becaon has been destroyed
+*/
+/obj/structure/beacon_generator/proc/destroyed_announcement(beacons_left = GLOB.infection_beacons.len)
+	sleep(80)
+	priority_announce("A beacon has been consumed by the infection, only [beacons_left] remain.","CentCom Biohazard Division", 'sound/misc/notice1.ogg')
```
Comment (PKPenguin321) [inline code/modules/antagonists/infection/beacon.dm]: Noticed this in the video, should probably read "only 1 remain***s***" when only one beacon remains

Comment (Whoneedspacee) [reply code/modules/antagonists/infection/beacon.dm]: ah crap i forgot about that thanks mate

**Comment by Whoneedspacee** [comment]
new video https://www.youtube.com/watch?v=D6zlblhr1Oc&feature=youtu.be showing off legendaries and looks a bit more polished

also holy crap please ignore that bug at the end god i messed up so bad

**Comment by bigfatbananacyclops** [comment]
please testmerge this on bagil

**Comment by 81Denton** [comment]
This is very ambitious, nice! I'm curious how this plays out as a testmerge, especially for regular crew.

**Comment by Akrilla** [comment]
Looks really great, music suits the whole theme too

**Comment by nemvar** [comment]
Silly request that I have because of that merge conflict. Don't add any more icons to items_and_weapons.dmi. It saves you and everyone else a headache.

**Comment by Whoneedspacee** [comment]
fair enough, there's so many defines and shit in the same place i figured why not for that, there's only like one thing in it anyways so i'll just move it out

**Comment by PKPenguin321** [comment]
e93

**Comment by Whoneedspacee** [comment]
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA

**Comment by actioninja** [comment]
image

**Comment by FuryMcFlurry** [comment]
OH hey, I did sprites in this one! Can't wait to play it one day.

**Comment by Bawhoppen** [comment]
As some other people previously mentioned, this gamemode should maybe have a different name. Infection makes it sound like it's some sort of zombie mode or somthing, and is also kind of generic-sounding for something which is more unique of a concept like this. 
My suggestion is to rename it to "Abberation," which I think will make it sound much more distinct and cool.

**Comment by Ordonis** [comment]
Infection mobs were not allowing ghosts to populate them when this got run yesterday.

**Comment by Whoneedspacee** [comment]
already fixed, will be in the next round of tests probably tommorow

**Comment by Whoneedspacee** [comment]
> As some other people previously mentioned, this gamemode should maybe have a different name. Infection makes it sound like it's some sort of zombie mode or somthing, and is also kind of generic-sounding for something which is more unique of a concept like this.
> My suggestion is to rename it to "Abberation," which I think will make it sound much more distinct and cool. (Doesn't "Abberation Commander" sound neat to you?)

it does but i don't really see how it fits the mode itself.

also jeez you really gonna make me rename all this shit in the code better be a damn good name haha.

**Comment by TerraGS** [comment]
I understand you want to wait for a meta to form, but this isn't a group of guildmates prepping for a raid. It's a server of impatient players who get tired of things quickly. My suggestion is start the infection off lower on the power scale and buff aspects accordingly, that way people are more willing to give it a chance rather than writing it off as bullshit.

**Comment by Whoneedspacee** [comment]
> I understand you want to wait for a meta to form, but this isn't a group of guildmates prepping for a raid. It's a server of impatient players who get tired of things quickly. My suggestion is start the infection off lower on the power scale and buff aspects accordingly, that way people are more willing to give it a chance rather than writing it off as bullshit.

I'm not against changing it between tests, not sure where you got that idea. It was already massively changed since the last test, and will be for this one. I put all the complaints I saw into a list and will try to address them.

**Comment by carlarctg** [comment]
>latejoin as doctor
>infection's taken over like one half of the station
>nobody is doing anything against it
>print a saw and slowly pick at infection spreading into the chemistry maintenance
>turret shocks me for 30 damage every second i hit the infection
>sentient slime sees me, runs at me outside of the infection and instantly crits me, with each hit giving me like 10 fire stacks 

>respawn as slime, agonizingly slow going from the core to the frontline
>dragging corpses takes literally forever, not bothering with that
>die to ERT (why do we need an ert to balance the round? what's the point of researching rnd, getting mutations, or mining for ore when you can just suicide and wait a few minutes to get a free pulse rifle?)
>no idea the closest node to the frontlines i need to respawn in, XY coords don't help at all
>waddling slowly to the frontlines
>i crash, ok
>rejoin, accidentally walk into perma because the infection makes me unable to view any landmarks
>need to leave
>come back, it crashed


not gonna lie, this sucked hard

**Comment by optimumtact** [comment]
@tgstation/commit-access closing in 24 hours unless someone merges

**Comment by SpaceManiac** [comment]
conflicts

**Comment by Whoneedspacee** [comment]
well there ya go its done for now, still a few changes that i'd like to make in the future like simplified controls for the commander but the basic gameplay should be fine and at least much more tolerable now.

**Comment by 81Denton** [comment]
@optimumtact This game mode has a proper concept and looks fun. I'm sadly unable to review it in the next 2-3 weeks.

**Comment by optimumtact** [comment]
then he can make a new pr in 2-3 weeks

**Comment by Whoneedspacee** [comment]
fair enough, gives me more time to add the few things i wanted to add anyways

**Comment by ExcessiveUseOfCobblestone** [comment]
@tralezab this looks similar to ur side-antag thing, perhaps u can share ideas?

**Review by ExcessiveUseOfCobblestone — COMMENTED** [review]
i'm like halfway lol

**Comment by Whoneedspacee** [comment]
:man_shrugging:

**Comment by ExcessiveUseOfCobblestone** [comment]
ill do bottom up soon lmfao

**Comment by Takov-Mala** [comment]
So, DGL had us test the mode out. I actually kinda liked it a bit but there's more than a few things that need fixing **big time**
The infection really doesn't need the bosses spawning, it's enough of a threat on its own. Seedlings are OP as fuck, pretty much instakills. Having to march through a region of respawning enemies full of tower defense turrets isn't exactly fun or feasible for the crew to manage. Gaining power via bodies means the infection just has to rush and kill some random fools to be large enough to have already won.

As an aside, with Cobbychem it's pretty much impossible as effectively there's very little healing that anyone is able to get once the first aid kits run out. Crew also pretty much are guaranteed to lose once a beacon is lost imo

**Comment by Whoneedspacee** [comment]
>no bosses

i can agree with this, but i like the legendary weapons they drop and i’m not sure how else to distribute them. definitely should be focusing more on the main slime gameplay rather than custom bosses at least.

>gaining power via bodies

is yes too strong, i do like the ghost roles so i’m not 100% sure how to bring them in still, but it’s too much yeah.

>lose once a beacon is lost

yeah i wasn’t really expecting the space dragon to destroy like half the stations research but i should have expected it. i guess i just imagined people moving departments but that’s kinda lame too. doing away with bosses and increasing damage being away from the infection for slimes would help a bit.

File: code/modules/antagonists/infection/infection/structures/node.dm [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	expansion_amount = 6
+
+/obj/structure/infection/node/mini/Initialize()
+	. = ..()
+	INVOKE_ASYNC(src, .proc/change_to_normal)
+
+/*
+	Changes the mini node to a normal infection after its life cycle has ended
+*/
+/obj/structure/infection/node/mini/proc/change_to_normal()
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/antagonists/infection/infection/structures/node.dm]: use addtimers/callbacks wtf

File: code/modules/antagonists/infection/infection/normalinfection.dm [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+		remove_atom_colour(FIXED_COLOUR_PRIORITY)
+
+/obj/structure/infection/process()
+	Life()
+
+/*
+	Life, pretty much the same as process
+*/
+/obj/structure/infection/proc/Life()
+	return
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/antagonists/infection/infection/normalinfection.dm]: why are you not just using process?

File: code/modules/antagonists/infection/infection/mobs/infection_mobs.dm [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	if(istype(src, /mob/living/simple_animal/hostile/infection/infectionspore/sentient))
+		return
+	if(exposed_temperature)
+		adjustFireLoss(CLAMP(0.01 * exposed_temperature, 1, 5))
+	else
+		adjustFireLoss(5)
+
+/mob/living/simple_animal/hostile/infection/CanPass(atom/movable/mover, turf/target)
+	if(istype(mover, /obj/structure/infection) || istype(mover, /obj/item/projectile/bullet/infection))
+		return 1
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/antagonists/infection/infection/mobs/infection_mobs.dm]: mom

**Comment by davethwave** [comment]
Just a thought on improving crewside of things. How about each time you kill a spore thing they drop a item that can permanently improve the crew member like +5 health or resistance or something like that so both sides improve with rare drops like some lavaland loot that would be useful so the commander and spores have to be a bit worried about every death as it improves the enemy side instead of rushing head on.

**Comment by ShizCalev** [comment]
conflict

**Comment by Whoneedspacee** [comment]
update comin out pretty soon will fix along w that

**Comment by Whoneedspacee** [comment]
i really need a break to just not think about anything tg related and relax. ss13 in general feels like its been a drain on my life for quite some time, pretty much every part of my life brings me some joy, and i just feel like crap whenever i participate here now.

thanks to everyone who cared, maybe i’ll come back and finish this at some point.

**Comment by PKPenguin321** [comment]
No worries. Come back whenever you feel like it man

**Comment by Akrilla** [comment]
mood

**Pull request closed without merging**